control/metrics.go

```diff
@@ -447,39 +290,134 @@ func (mc *metricCatalog) RmUnloadedPluginMetrics(lp *loadedPlugin) {
 	mc.mutex.Lock()
 	defer mc.mutex.Unlock()
-	return mc.get(ns.Strings(), version)
+
+	var ns core.Namespace
+
+	catalogedmt, err := mc.tree.GetMetric(requested.Strings(), version)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"_module": "control",
+			"_file":   "metrics.go,",
+			"_block":  "get-metric",
+			"error":   err,
+		}).Error("error getting metric")
+		return nil, err
+	}
+
+	ns = catalogedmt.Namespace()
+
+	if isDynamic, _ := ns.IsDynamic(); isDynamic {
+		// specify instance of dynamicMetric
+		if ns.String() != requested.String() {
+			ns = specifyInstanceOfDynamicMetric(ns, requested)
+		}
+	}
+
+	returnedmt := &metricType{
+		Plugin:             catalogedmt.Plugin,
+		namespace:          ns,
+		version:            catalogedmt.Version(),
+		lastAdvertisedTime: catalogedmt.LastAdvertisedTime(),
+		tags:               catalogedmt.Tags(),
+		policy:             catalogedmt.Plugin.ConfigPolicy.Get(catalogedmt.Namespace().Strings()),
+		config:             catalogedmt.Config(),
+		unit:               catalogedmt.Unit(),
+		description:        catalogedmt.Description(),
+		subscriptions:      catalogedmt.SubscriptionCount(),
+	}
+	return returnedmt, nil
+}
+
+// GetMetrics retrieves all metrics which fulfill a given requested namespace and version.
+// If provided a version of -1 the latest plugin will be returned.
+func (mc *metricCatalog) GetMetrics(requested core.Namespace, version int) ([]*metricType, error) {
+	mc.mutex.Lock()
+	defer mc.mutex.Unlock()
+
+	returnedmts := []*metricType{}
+
+	// resolve queried tuples in metric namespace
+	requestedNss := findTuplesMatches(requested)
+	for _, rns := range requestedNss {
+		catalogedmts, err := mc.tree.GetMetrics(rns.Strings(), version)
+		if err != nil {
+			log.WithFields(log.Fields{
+				"_module": "control",
+				"_file":   "metrics.go,",
+				"_block":  "get-metrics",
+				"error":   err,
+			}).Error("error getting metric")
+			return nil, err
+		}
+		for _, catalogedmt := range catalogedmts {
+			ns := catalogedmt.Namespace()
+
+			if isDynamic, _ := ns.IsDynamic(); isDynamic {
+				// specify instance of dynamicMetric when cataloged namespace is different that requested
+				if ns.String() != rns.String() {
+					ns = specifyInstanceOfDynamicMetric(ns, rns)
+				}
+			}
+
+			returnedmt := &metricType{
+				Plugin:             catalogedmt.Plugin,
+				namespace:          ns,
+				version:            catalogedmt.Version(),
+				lastAdvertisedTime: catalogedmt.LastAdvertisedTime(),
+				tags:               catalogedmt.Tags(),
+				policy:             catalogedmt.Plugin.ConfigPolicy.Get(catalogedmt.Namespace().Strings()),
+				config:             catalogedmt.Config(),
+				unit:               catalogedmt.Unit(),
+				description:        catalogedmt.Description(),
+				subscriptions:      catalogedmt.SubscriptionCount(),
+			}
+			returnedmts = appendIfUnique(returnedmts, returnedmt)
```

Comment:
This never happens, **so definitely `appendIfUnique` can be removed.** Probably I suggested on appendIfMissing which is use for metrics' keys. Thanks for catching that.

The only possible when it could have sense is requesting the same items in tuple (as "/intel/mock/(host0;host0)/baz"), but I think it's ok for that returning two metrics - user should notice that duplication and make correction on his own.